sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -501,6 +501,10 @@ object BooleanSimplification extends Rule[LogicalPlan] with PredicateHelper {
         case LessThanOrEqual(l, r) => GreaterThan(l, r)
         // not(not(e))  =>  e
         case Not(e) => e
+        // De Morgan's law: !(a || b) => !a && !b
```

Comment:
As mentioned in https://github.com/apache/spark/pull/8200#discussion_r37083539, I tend not to have this one in `BooleanSimplification`. I'd prefer not having CNF conversion coupled with `BooleanSimplification`, and do transformations related to `Not` in the CNF part.